### DISCUSSION TITLE
Heterogeneous graph learning and explanation with custom dataset

### DISCUSSION THREAD

**Discussion by SamiDalhoumi** [pyg-team/pytorch_geometric#7563] [Q&A]
Hi! I'm relatively new to PyG and I'm trying to build and explain an heterogeneous graph dataset using my own data. I managed to create the heterogeneous dataset and train a model but I'm stuck in graphs explanation step. I have some questions regarding heterogeneous graph learning using PyG:

- Why should we convert directed heterogeneous graphs to undirected graphs before learning graph models on them?

- I'm stuck in graph explanation step, I tried out different things (node-level, graph-level regression, etc.) but still getting error messages.
![image](https://github.com/pyg-team/pytorch_geometric/assets/79154982/53c314ee-1891-4f25-b1a8-5f7cfba3a672)

Attached a notebook and a dummy dataset to reproduce the example.

Thank you for help!
[test_hybrid_gnn.zip](https://github.com/pyg-team/pytorch_geometric/files/11724523/test_hybrid_gnn.zip)

**Comment by wsad1** [comment]
> Why should we convert directed heterogeneous graphs to undirected graphs before learning graph models on them?

This isn't required. Did you face some issue when you didn't do this?

> I'm stuck in graph explanation step, I tried out different things (node-level, graph-level regression, etc.) but still getting error messages.

I looked at your notebook. What error did you get while doing node-level explanation. Could you paste the full stack trace here.

**Reply by SamiDalhoumi** [reply]
Thanks for reply!

Regarding first issue, when I comment these lines in graph dataset creation part and create an undirected graphs dataset:

![image](https://github.com/pyg-team/pytorch_geometric/assets/79154982/347e3977-bf08-40fe-bc28-f2be7b8a413b)

Then in model training and testing part, when I remove the try / except

![image](https://github.com/pyg-team/pytorch_geometric/assets/79154982/b5169f2f-92b0-4b35-a885-b0a97b67e7f1)

I get this error:
--------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[18], line 39
     34     return avg_loss / len(data_set)
     37 for epoch in range(1, 10): # just few epochs for testing
---> 39     train()
     40     print('===========Train eval===========')
     41     train_loss = test(train_set)

Cell In[18], line 14, in train()
     10     model.train()
     12     for i, data in enumerate(train_set): 
     13 #         try:
---> 14         out = model(data.x_dict, data.edge_index_dict)
     15         loss = criterion(out, data.y)
     16         loss.backward()

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\torch\nn\modules\module.py:1501, in Module._call_impl(self, *args, **kwargs)
   1496 # If we don't have any hooks, we want to skip the rest of the logic in
   1497 # this function, and just call forward.
   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1499         or _global_backward_pre_hooks or _global_backward_hooks
   1500         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1501     return forward_call(*args, **kwargs)
   1502 # Do not call functions when jit is used
   1503 full_backward_hooks, non_full_backward_hooks = [], []

Cell In[17], line 33, in HGT.forward(self, x_dict, edge_index_dict)
     30     x_dict[node_type] = self.lin_dict[node_type](x).relu_()
     32 for conv in self.convs:
---> 33     x_dict = conv(x_dict, edge_index_dict)
     35 return self.lin(x_dict['S10'])

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\torch\nn\modules\module.py:1501, in Module._call_impl(self, *args, **kwargs)
   1496 # If we don't have any hooks, we want to skip the rest of the logic in
   1497 # this function, and just call forward.
   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1499         or _global_backward_pre_hooks or _global_backward_hooks
   1500         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1501     return forward_call(*args, **kwargs)
   1502 # Do not call functions when jit is used
   1503 full_backward_hooks, non_full_backward_hooks = [], []

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\torch_geometric\nn\conv\hgt_conv.py:155, in HGTConv.forward(self, x_dict, edge_index_dict)
    153 # Iterate over node-types:
    154 for node_type, x in x_dict.items():
--> 155     k_dict[node_type] = self.k_lin[node_type](x).view(-1, H, D)
    156     q_dict[node_type] = self.q_lin[node_type](x).view(-1, H, D)
    157     v_dict[node_type] = self.v_lin[node_type](x).view(-1, H, D)

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\torch\nn\modules\module.py:1501, in Module._call_impl(self, *args, **kwargs)
   1496 # If we don't have any hooks, we want to skip the rest of the logic in
   1497 # this function, and just call forward.
   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1499         or _global_backward_pre_hooks or _global_backward_hooks
   1500         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1501     return forward_call(*args, **kwargs)
   1502 # Do not call functions when jit is used
   1503 full_backward_hooks, non_full_backward_hooks = [], []

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\torch_geometric\nn\dense\linear.py:132, in Linear.forward(self, x)
    127 def forward(self, x: Tensor) -> Tensor:
    128     r"""
    129     Args:
    130         x (torch.Tensor): The input features.
    131     """
--> 132     return F.linear(x, self.weight, self.bias)

TypeError: linear(): argument 'input' (position 1) must be Tensor, not NoneType

==================================

Regarding second issue, when trying to do graph-level or node-level explanation, I get this error message:

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[18], line 14
      1 explainer = Explainer(
      2     model, 
      3     algorithm=CaptumExplainer('IntegratedGradients'),
   (...)
     11     ),
     12 )
---> 14 hetero_explanation = explainer(
     15     x=dataset[0].x_dict,
     16     edge_index=dataset[0].edge_index_dict,
     17     target=dataset[0],
     18 )
     19 print(hetero_explanation.edge_mask_dict)
     20 print(hetero_explanation.node_mask_dict)

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\torch_geometric\explain\explainer.py:198, in Explainer.__call__(self, x, edge_index, target, index, **kwargs)
    195 training = self.model.training
    196 self.model.eval()
--> 198 explanation = self.algorithm(
    199     self.model,
    200     x,
    201     edge_index,
    202     target=target,
    203     index=index,
    204     **kwargs,
    205 )
    207 self.model.train(training)
    209 # Add explainer objectives to the `Explanation` object:

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\torch\nn\modules\module.py:1501, in Module._call_impl(self, *args, **kwargs)
   1496 # If we don't have any hooks, we want to skip the rest of the logic in
   1497 # this function, and just call forward.
   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1499         or _global_backward_pre_hooks or _global_backward_hooks
   1500         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1501     return forward_call(*args, **kwargs)
   1502 # Do not call functions when jit is used
   1503 full_backward_hooks, non_full_backward_hooks = [], []

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\torch_geometric\explain\algorithm\captum_explainer.py:162, in CaptumExplainer.forward(self, model, x, edge_index, target, index, **kwargs)
    159 else:
    160     target = target[index]
--> 162 attributions = self.attribution_method.attribute(
    163     inputs=inputs,
    164     target=target,
    165     additional_forward_args=add_forward_args,
    166     **self.kwargs,
    167 )
    169 node_mask, edge_mask = convert_captum_output(
    170     attributions,
    171     mask_type,
    172     metadata,
    173 )
    175 if not isinstance(x, dict):

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\captum\log\__init__.py:42, in log_usage.<locals>._log_usage.<locals>.wrapper(*args, **kwargs)
     40 @wraps(func)
     41 def wrapper(*args, **kwargs):
---> 42     return func(*args, **kwargs)

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\captum\attr\_core\integrated_gradients.py:274, in IntegratedGradients.attribute(self, inputs, baselines, target, additional_forward_args, n_steps, method, internal_batch_size, return_convergence_delta)
    272 if internal_batch_size is not None:
    273     num_examples = inputs[0].shape[0]
--> 274     attributions = _batch_attribution(
    275         self,
    276         num_examples,
    277         internal_batch_size,
    278         n_steps,
    279         inputs=inputs,
    280         baselines=baselines,
    281         target=target,
    282         additional_forward_args=additional_forward_args,
    283         method=method,
    284     )
    285 else:
    286     attributions = self._attribute(
    287         inputs=inputs,
    288         baselines=baselines,
   (...)
    292         method=method,
    293     )

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\captum\attr\_utils\batching.py:78, in _batch_attribution(attr_method, num_examples, internal_batch_size, n_steps, include_endpoint, **kwargs)
     76 step_sizes = full_step_sizes[start_step:end_step]
     77 alphas = full_alphas[start_step:end_step]
---> 78 current_attr = attr_method._attribute(
     79     **kwargs, n_steps=batch_steps, step_sizes_and_alphas=(step_sizes, alphas)
     80 )
     82 if total_attr is None:
     83     total_attr = current_attr

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\captum\attr\_core\integrated_gradients.py:351, in IntegratedGradients._attribute(self, inputs, baselines, target, additional_forward_args, n_steps, method, step_sizes_and_alphas)
    348 expanded_target = _expand_target(target, n_steps)
    350 # grads: dim -> (bsz * #steps x inputs[0].shape[1:], ...)
--> 351 grads = self.gradient_func(
    352     forward_fn=self.forward_func,
    353     inputs=scaled_features_tpl,
    354     target_ind=expanded_target,
    355     additional_forward_args=input_additional_args,
    356 )
    358 # flattening grads so that we can multilpy it with step-size
    359 # calling contiguous to avoid `memory whole` problems
    360 scaled_grads = [
    361     grad.contiguous().view(n_steps, -1)
    362     * torch.tensor(step_sizes).view(n_steps, 1).to(grad.device)
    363     for grad in grads
    364 ]

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\captum\_utils\gradient.py:112, in compute_gradients(forward_fn, inputs, target_ind, additional_forward_args)
     94 r"""
     95 Computes gradients of the output with respect to inputs for an
     96 arbitrary forward function.
   (...)
    108                 arguments) if no additional arguments are required
    109 """
    110 with torch.autograd.set_grad_enabled(True):
    111     # runs forward pass
--> 112     outputs = _run_forward(forward_fn, inputs, target_ind, additional_forward_args)
    113     assert outputs[0].numel() == 1, (
    114         "Target not provided when necessary, cannot"
    115         " take gradient with respect to multiple outputs."
    116     )
    117     # torch.unbind(forward_out) is a list of scalar tensor tuples and
    118     # contains batch_size * #steps elements

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\captum\_utils\common.py:482, in _run_forward(forward_func, inputs, target, additional_forward_args)
    479 inputs = _format_inputs(inputs)
    480 additional_forward_args = _format_additional_forward_args(additional_forward_args)
--> 482 output = forward_func(
    483     *(*inputs, *additional_forward_args)
    484     if additional_forward_args is not None
    485     else inputs
    486 )
    487 return _select_targets(output, target)

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\torch\nn\modules\module.py:1501, in Module._call_impl(self, *args, **kwargs)
   1496 # If we don't have any hooks, we want to skip the rest of the logic in
   1497 # this function, and just call forward.
   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1499         or _global_backward_pre_hooks or _global_backward_hooks
   1500         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1501     return forward_call(*args, **kwargs)
   1502 # Do not call functions when jit is used
   1503 full_backward_hooks, non_full_backward_hooks = [], []

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\torch_geometric\explain\algorithm\captum.py:152, in CaptumHeteroModel.forward(self, *args)
    149     x = self.model(x_dict, edge_index_dict,
    150                    *args[-len_remaining_args:])
    151 else:
--> 152     x = self.model(x_dict, edge_index_dict)
    154 if self.mask_type.with_edge:
    155     clear_masks(self.model)

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\torch\nn\modules\module.py:1501, in Module._call_impl(self, *args, **kwargs)
   1496 # If we don't have any hooks, we want to skip the rest of the logic in
   1497 # this function, and just call forward.
   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1499         or _global_backward_pre_hooks or _global_backward_hooks
   1500         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1501     return forward_call(*args, **kwargs)
   1502 # Do not call functions when jit is used
   1503 full_backward_hooks, non_full_backward_hooks = [], []

Cell In[16], line 30, in HGT.forward(self, x_dict, edge_index_dict)
     27 def forward(self, x_dict, edge_index_dict):
     29     for node_type, x in x_dict.items():
---> 30         x_dict[node_type] = self.lin_dict[node_type](x).relu_()
     32     for conv in self.convs:
     33         x_dict = conv(x_dict, edge_index_dict)

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\torch\nn\modules\module.py:1501, in Module._call_impl(self, *args, **kwargs)
   1496 # If we don't have any hooks, we want to skip the rest of the logic in
   1497 # this function, and just call forward.
   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1499         or _global_backward_pre_hooks or _global_backward_hooks
   1500         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1501     return forward_call(*args, **kwargs)
   1502 # Do not call functions when jit is used
   1503 full_backward_hooks, non_full_backward_hooks = [], []

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\torch_geometric\nn\dense\linear.py:132, in Linear.forward(self, x)
    127 def forward(self, x: Tensor) -> Tensor:
    128     r"""
    129     Args:
    130         x (torch.Tensor): The input features.
    131     """
--> 132     return F.linear(x, self.weight, self.bias)

RuntimeError: mat1 and mat2 shapes cannot be multiplied (1x32 and 10x32)

I use torch-geometric 2.3.1

**Reply by rusty1s** [reply]
I think I found the issue. It looks like you write to `x_dict` in-place, which causes trouble since `captum` repeatedely calls your model (but now with wrong inputs). Please try to change
```python
for node_type, x in x_dict.items():
    x_dict[node_type] = self.lin_dict[node_type](x).relu_()
```
to
```python
x_dict = {key: self.lin_dict[key](x).relu() for key, x in x_dict.items()}

**Reply by SamiDalhoumi** [reply]
Still not working. I've got this error when running model explanation cell. 
You use the same code for model explanation?

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[20], line 14
      1 explainer = Explainer(
      2     model, 
      3     algorithm=CaptumExplainer('IntegratedGradients'),
   (...)
     11     ),
     12 )
---> 14 hetero_explanation = explainer(
     15     x=dataset[0].x_dict,
     16     edge_index=dataset[0].edge_index_dict,
     17     target=dataset[0],
     18 )
     20 # explainer = Explainer(
     21 #     model,  # It is assumed that model outputs a single tensor.
     22 #     algorithm=CaptumExplainer('IntegratedGradients'),
   (...)
     37 #     dataset[0].edge_index_dict
     38 # )
     40 print(hetero_explanation.edge_mask_dict)

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\torch_geometric\explain\explainer.py:198, in Explainer.__call__(self, x, edge_index, target, index, **kwargs)
    195 training = self.model.training
    196 self.model.eval()
--> 198 explanation = self.algorithm(
    199     self.model,
    200     x,
    201     edge_index,
    202     target=target,
    203     index=index,
    204     **kwargs,
    205 )
    207 self.model.train(training)
    209 # Add explainer objectives to the `Explanation` object:

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\torch\nn\modules\module.py:1501, in Module._call_impl(self, *args, **kwargs)
   1496 # If we don't have any hooks, we want to skip the rest of the logic in
   1497 # this function, and just call forward.
   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1499         or _global_backward_pre_hooks or _global_backward_hooks
   1500         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1501     return forward_call(*args, **kwargs)
   1502 # Do not call functions when jit is used
   1503 full_backward_hooks, non_full_backward_hooks = [], []

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\torch_geometric\explain\algorithm\captum_explainer.py:162, in CaptumExplainer.forward(self, model, x, edge_index, target, index, **kwargs)
    159 else:
    160     target = target[index]
--> 162 attributions = self.attribution_method.attribute(
    163     inputs=inputs,
    164     target=target,
    165     additional_forward_args=add_forward_args,
    166     **self.kwargs,
    167 )
    169 node_mask, edge_mask = convert_captum_output(
    170     attributions,
    171     mask_type,
    172     metadata,
    173 )
    175 if not isinstance(x, dict):

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\captum\log\__init__.py:42, in log_usage.<locals>._log_usage.<locals>.wrapper(*args, **kwargs)
     40 @wraps(func)
     41 def wrapper(*args, **kwargs):
---> 42     return func(*args, **kwargs)

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\captum\attr\_core\integrated_gradients.py:274, in IntegratedGradients.attribute(self, inputs, baselines, target, additional_forward_args, n_steps, method, internal_batch_size, return_convergence_delta)
    272 if internal_batch_size is not None:
    273     num_examples = inputs[0].shape[0]
--> 274     attributions = _batch_attribution(
    275         self,
    276         num_examples,
    277         internal_batch_size,
    278         n_steps,
    279         inputs=inputs,
    280         baselines=baselines,
    281         target=target,
    282         additional_forward_args=additional_forward_args,
    283         method=method,
    284     )
    285 else:
    286     attributions = self._attribute(
    287         inputs=inputs,
    288         baselines=baselines,
   (...)
    292         method=method,
    293     )

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\captum\attr\_utils\batching.py:78, in _batch_attribution(attr_method, num_examples, internal_batch_size, n_steps, include_endpoint, **kwargs)
     76 step_sizes = full_step_sizes[start_step:end_step]
     77 alphas = full_alphas[start_step:end_step]
---> 78 current_attr = attr_method._attribute(
     79     **kwargs, n_steps=batch_steps, step_sizes_and_alphas=(step_sizes, alphas)
     80 )
     82 if total_attr is None:
     83     total_attr = current_attr

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\captum\attr\_core\integrated_gradients.py:351, in IntegratedGradients._attribute(self, inputs, baselines, target, additional_forward_args, n_steps, method, step_sizes_and_alphas)
    348 expanded_target = _expand_target(target, n_steps)
    350 # grads: dim -> (bsz * #steps x inputs[0].shape[1:], ...)
--> 351 grads = self.gradient_func(
    352     forward_fn=self.forward_func,
    353     inputs=scaled_features_tpl,
    354     target_ind=expanded_target,
    355     additional_forward_args=input_additional_args,
    356 )
    358 # flattening grads so that we can multilpy it with step-size
    359 # calling contiguous to avoid `memory whole` problems
    360 scaled_grads = [
    361     grad.contiguous().view(n_steps, -1)
    362     * torch.tensor(step_sizes).view(n_steps, 1).to(grad.device)
    363     for grad in grads
    364 ]

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\captum\_utils\gradient.py:119, in compute_gradients(forward_fn, inputs, target_ind, additional_forward_args)
    113     assert outputs[0].numel() == 1, (
    114         "Target not provided when necessary, cannot"
    115         " take gradient with respect to multiple outputs."
    116     )
    117     # torch.unbind(forward_out) is a list of scalar tensor tuples and
    118     # contains batch_size * #steps elements
--> 119     grads = torch.autograd.grad(torch.unbind(outputs), inputs)
    120 return grads

File ~\Anaconda3\envs\pyg_gnn\Lib\site-packages\torch\autograd\__init__.py:303, in grad(outputs, inputs, grad_outputs, retain_graph, create_graph, only_inputs, allow_unused, is_grads_batched)
    301     return _vmap_internals._vmap(vjp, 0, 0, allow_none_pass_through=True)(grad_outputs_)
    302 else:
--> 303     return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
    304         t_outputs, grad_outputs_, retain_graph, create_graph, t_inputs,
    305         allow_unused, accumulate_grad=False)

RuntimeError: One of the differentiated Tensors appears to not have been used in the graph. Set allow_unused=True if this is the desired behavior.

**Reply by rusty1s** [reply]
Ok, so this is a different issue. Can you investigate if there exists certain parameters that won't get upgraded? I assume you may have some node types in your graph that are not reachable?

**Reply by SamiDalhoumi** [reply]
There is no isolated nodes in my graphs

**Reply by rusty1s** [reply]
Yes, but can you check during training whether there exists parameters with no gradients? For example, something like
```python
loss.backward()
for param in model.parameters():
    assert param.grad is not None
    assert param.grad.abs().sum() > 0

**Reply by SamiDalhoumi** [reply]
It seems that there are parameters with no gradients! Is it a data or model issue? I just copied model structure from PyG documentation

**Reply by rusty1s** [reply]
Sorry for late reply. Do you know which node types/edge types do not have any gradients? What happens if you remove some types which may not be reachable within k-hop subgraphs, e.g., via `del data[node_type]`?

**Reply by SamiDalhoumi** [reply]
Below an example of the structure of one graph in my dataset. I convert graphs to undirected because it seems that heterogeneous GNNs don't work with directed datasets.
![image](https://github.com/pyg-team/pytorch_geometric/assets/79154982/f4ff1ab2-3071-4e43-8f86-d9e1ee33dd79)

I tried to create graphs using only nodes types (S3, S4, S5), then (S6, S7, S8) and I've got the same issue for both cases.

The only heterogeneous graphs creation method that works for me is using existing heterogeneous operators, but I get parameters with no gradients: https://pytorch-geometric.readthedocs.io/en/latest/tutorial/heterogeneous.html
